refactor: Remove custom hit testing in usePress

devongovett opened the pull request:
Related to #1720, #2506, #5290

Back in #62 we implemented our own hit testing in usePress in order to work around a bug in pointer events in iOS 13. Now that this version is long gone, we can probably remove this and use the `onPointerLeave` and `onPointerEnter` events instead. This depends on [releasePointerCapture](https://developer.mozilla.org/en-US/docs/Web/API/Element/releasePointerCapture) being called during `onPointerDown` on touch devices, so that the original target doesn't capture the pointer and cause the browser to retarget all events to it even outside the hit area.

devongovett commented on packages/@react-aria/interactions/src/usePress.ts:
```diff
@@ -433,7 +433,13 @@ export function usePress(props: PressHookProps): PressResult {
+          // Release pointer capture so that touch interactions can leave the original target.
+          // This enables onPointerLeave and onPointerEnter to fire.
+          let target = e.target as Element;
+          if ('releasePointerCapture' in target) {
```
not supported in jsdom

devongovett commented on packages/@react-aria/interactions/src/usePress.ts:
```diff
@@ -467,27 +473,20 @@ export function usePress(props: PressHookProps): PressResult {
-        // Safari on iOS sometimes fires pointerup events, even
-        // when the touch isn't over the target, so double check.
-        if (e.button === 0 && isOverTarget(e, e.currentTarget)) {
+        if (e.button === 0) {
```
I am not certain yet if this bug still exists in Safari. This was added in #335. Will need to test.

snowystinger reviewed (approved):
Excited to get this into testing!

devongovett commented:
Updated to address two issues found in testing:

> On touch (ipad), open any sub menu and press down on a submenu item, without lifting, slide off the submenu and release. Menus all close now. The close behavior exists already with mouse.

This only happens in RSP v3, not in RAC, because we recently changed the DOM structure to no longer be nested there. In RSP we have an extra `useInteractOutside` in Menu, but it only checks if the interaction was outside the menu itself not any submenus. It checks that during `onInteractOutside` but not in the pointer down event. I moved this to the root menu trigger instead so submenus should be included.

> On ipad, starting a touch event on one item and then dragging and releasing of a different item selects that item. Doesn't happen on Android and doesn't happen if the dropdown is scrollable. Didn't happen on main previously, same issue as row 7. Also occurs for Menu (DL)

This works as intended with mouse (matching native behavior), but it is weird with touch since the user might be trying to scroll. Updated this logic to only select during `onPressUp` if `pointerType` is mouse, otherwise wait for `onPress` which requires pressing down on the same target your release over.

snowystinger reviewed (approved):
submenus appear to be working correctly in s1 and s2

LFDanLu reviewed (approved):
LGTM, verified the menu/pickers/etc don't auto close on touch when attempting to scroll. Also verified submenus don't close when dragging off of them in touch anymore
